### PR TITLE
bessctl: expose deadend packets per module

### PR DESCRIPTION
**Pull request by melvinw**

### CODE REVIEW

File: core/module.h [hunk 1]
```diff
@@ -441,6 +443,7 @@ class alignas(64) Module {
 
   std::vector<bess::IGate *> igates_;
   std::vector<bess::OGate *> ogates_;
+  uint64_t deadends_;
```
Comment (nrdmao) [inline core/module.h]: Shouldn't you have a counter per worker as multiple workers may be running through the module at the same time. As such deadends() would return the sum of the counters for each worker.

Comment (melvinw) [reply core/module.h]: good catch! adjusted

File: core/module.h [hunk 2]
```diff
@@ -495,13 +506,15 @@ inline void Module::RunChooseModule(Context *ctx, gate_idx_t ogate_idx,
...
   if (unlikely(ogate_idx >= ogates_.size())) {
+    deadends_[ctx->wid]++;
     deadend(ctx, batch);
     return;
   }
 
   ogate = ogates_[ogate_idx];
 
   if (unlikely(!ogate)) {
+    deadends_[ctx->wid]++;
```
Comment (sangjinhan) [inline core/module.h]: ditto

File: core/module.h [hunk 3]
```diff
@@ -327,6 +329,14 @@ class alignas(64) Module {
 
   const std::vector<bess::OGate *> &ogates() const { return ogates_; }
 
+  uint64_t deadends() const {
+    uint64_t ret = 0;
```
Comment (sangjinhan) [inline core/module.h]: if you like: `return std::accumulate(deadends_, deadends_ + kMaxWorkers, 0);`

File: core/module.h [hunk 4]
```diff
@@ -441,6 +451,7 @@ class alignas(64) Module {
 
   std::vector<bess::IGate *> igates_;
   std::vector<bess::OGate *> ogates_;
+  uint64_t deadends_[Worker::kMaxWorkers];
```
Comment (sangjinhan) [inline core/module.h]: if you like: `std::array<uint64_t, kMaxWorkers> deadends_`

File: core/module.h (hunk 3 above)
Comment (sangjinhan) [reply core/module.h]: (or `.begin()` and `.end()` for an `std::array`)

**Comment by melvinw** [comment]
Thanks @sangjinhan! Updated

File: core/module.h [hunk 5]
```diff
@@ -327,6 +331,10 @@ class alignas(64) Module {
 
   const std::vector<bess::OGate *> &ogates() const { return ogates_; }
 
+  uint64_t deadends() const {
```
Comment (shinae-woo) [inline core/module.h]: I guess we want per-worker statistics here either array or map. The caller (`bessctl.cc`) able to accumulation if required.

Comment (melvinw) [reply core/module.h]: In my head `deadends_` is just an array to make the counter thread-safe-- i.e., it's just an implementation detail. I only ever intended to expose a single counter to the user here

**Comment by melvinw** [comment]
Nice catch @shinae-woo! Pushed changes to `RunSplit()` and `ResetActiveWorkers()`

**Comment by shinae-woo** [comment]
Thank you! LGTM for the part I commented on. @sangjinhan Do you wanna take a look again and merge this?

**Pull request merged**